RFC: Create trait for easier route generation

alexander-schranz opened the pull request:
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| License | MIT

#### What's in this PR?

Create trait for easier route generation.

#### Why?

Currently to create just a list with a add and edit form is really hard without copying that from another bundle as its hard to remind how you need to create this routes.

#### Example Usage

~~~php
    public function getRoutes(): array
    {
        $locales = ...

        return array_merge(
            $this->createBasicRoutes('events', '/events', $locales)
            $this->createBasicRoutes('notices', '/notices')
            $this->createListRoute('logs', '/logs')
        );
    }

    public function getJsBundleName()
    {
        return 'app';
    }
~~~

wachterjohannes commented:
i like the idea - although we should think about writing Builders for the routes (would make the writing a bit more readable). 

@alexander-schranz could you maybe refactor one admin class to see the trait in action?

alexander-schranz commented:
fixed version I use in current project:

Its updated to the actual sulu version and moved the backRoute be on the tabs resource instead of a single tab.

```php
declare(strict_types=1);

namespace App\Admin;

use Sulu\Bundle\AdminBundle\Admin\Routing\Route;

trait RouteListFormTrait
{
    abstract function getPrefix();
    
    private function createRoutes(string $resourceKey, string $path, array $locales = [])
    {
        $prefix = $this->getPrefix();
        $localePrefix = '';

        if (\count($locales)) {
            $localePrefix = '/:locale';
        }

        $formToolbarActions = [
            'sulu_admin.save',
            'sulu_admin.delete',
        ];

        return [
            (new Route($prefix . '.' . $resourceKey . '.datagrid', $path . $localePrefix, 'sulu_admin.datagrid'))
                ->addOption('title', $prefix . '.' . $resourceKey)
                ->addOption('adapters', ['table'])
                ->addOption('resourceKey', $resourceKey)
                ->addOption('addRoute', $prefix . '.' . $resourceKey . '.add.detail')
                ->addOption('editRoute', $prefix . '.' . $resourceKey . '.edit.detail')
                ->addOption('locales', $locales)
                ->addAttributeDefault('locale', $locales[0] ?: ''),
            // Add tabs
            (new Route($prefix . '.' . $resourceKey . '.add', $path . $localePrefix . '/add', 'sulu_admin.resource_tabs'))
                ->addOption('resourceKey', $resourceKey)
                ->addOption('toolbarActions', $formToolbarActions)
                ->addOption('backRoute', $prefix . '.' . $resourceKey . '.datagrid')
                ->addOption('locales', $locales)
                ->addAttributeDefault('locale', $locales[0] ?: ''),
            // Add detail
            (new Route($prefix . '.' . $resourceKey . '.add.detail', '/details', 'sulu_admin.form'))
                ->addOption('tabTitle', $prefix . '.details')
                ->addOption('editRoute', $prefix . '.' . $resourceKey . '.edit.detail')
                ->addOption('toolbarActions', $formToolbarActions)
                ->setParent($prefix . '.' . $resourceKey . '.add'),
            // Edit tabs
            (new Route($prefix . '.' . $resourceKey . '.edit', $path . $localePrefix . '/:id', 'sulu_admin.resource_tabs'))
                ->addOption('resourceKey', $resourceKey)
                ->addOption('toolbarActions', $formToolbarActions)
                ->addOption('backRoute', $prefix . '.' . $resourceKey . '.datagrid')
                ->addOption('locales', $locales)
                ->addAttributeDefault('locale', $locales[0] ?: ''),
            // Edit detail
            (new Route($prefix . '.' . $resourceKey . '.edit.detail', '/details', 'sulu_admin.form'))
                ->addOption('tabTitle', $prefix . '.details')
                ->addOption('toolbarActions', $formToolbarActions)
                ->setParent($prefix . '.' . $resourceKey . '.edit'),
        ];
    }
}
```
